internal/controller/apiextensions/composite/composition_ptf.go

```diff
@@ -0,0 +1,982 @@
+	// already wrapped? Or just do away with unstructured.NewClient completely?
+	kube = unstructured.NewClient(kube)
+
+	f := NewSecretConnectionDetailsFetcher(kube)
+
+	c := &PTFComposer{
+		client: resource.ClientApplicator{Client: kube, Applicator: resource.NewAPIUpdatingApplicator(kube)},
+
+		composite: ptfComposite{
+			ConnectionDetailsFetcher: f,
+			ComposedResourceGetter:   NewExistingComposedResourceGetter(kube, f),
+			ComposedResourceDeleter:  NewUndesiredComposedResourceDeleter(kube),
+			ComposedResourceObserver: ComposedResourceObserverChain{
+				NewConnectionDetailsObserver(ConnectionDetailsExtractorFn(ExtractConnectionDetails)),
+				NewReadinessObserver(ReadinessCheckerFn(IsReady)),
+			},
+		},
+		composition: ptfComposition{
+			PatchAndTransformer:    NewXRCDPatchAndTransformer(RendererFn(RenderComposite), NewAPIDryRunRenderer(kube)),
+			FunctionPipelineRunner: NewFunctionPipeline(ContainerFunctionRunnerFn(RunFunction)),
+		},
+	}
+
+	for _, fn := range o {
+		fn(c)
+	}
+
+	return c
+}
+
+// PTFCompositionState is used throughout the PTFComposer to track its state.
+type PTFCompositionState struct {
+	Composite         resource.Composite
+	ConnectionDetails managed.ConnectionDetails
+	ComposedResources ComposedResourceStates
+	Events            []event.Event
+}
+
+// Compose resources using both either the Patch & Transform style resources
+// array, the functions array, or both.
+func (c *PTFComposer) Compose(ctx context.Context, xr resource.Composite, req CompositionRequest) (CompositionResult, error) { //nolint:gocyclo // We probably don't want any further abstraction for the sake of reduced complexity.
+	xc, err := c.composite.FetchConnection(ctx, xr)
+	if err != nil {
+		return CompositionResult{}, errors.Wrap(err, errFetchXRConnectionDetails)
+	}
+
+	cds, err := c.composite.GetComposedResources(ctx, xr)
+	if err != nil {
+		return CompositionResult{}, errors.Wrap(err, errGetExistingCDs)
+	}
+
+	state := &PTFCompositionState{
+		Composite:         xr,
+		ConnectionDetails: xc,
+		ComposedResources: cds,
+		Events:            make([]event.Event, 0),
+	}
+
+	// Build observed state to be passed to our Composition Function pipeline.
+	// Doing this before we patch and transform ensures we report the state we
+	// actually observed before we made any mutations.
+	o, err := FunctionIOObserved(state)
+	if err != nil {
+		return CompositionResult{}, errors.Wrap(err, errBuildFunctionIOObserved)
+	}
+
+	// Run P&T logic, updating the composition state accordingly.
+	if err := c.composition.PatchAndTransform(ctx, req, state); err != nil {
+		return CompositionResult{}, errors.Wrap(err, errPatchAndTransform)
+	}
+
+	// Build the initial desired state to be passed to our Composition Function
+	// pipeline. It's expected that each function in the pipeline will mutate
+	// this state. It includes any desired state accumulated by the P&T logic.
+	d, err := FunctionIODesired(state)
```

Comment:
During my experiments, I found out that the list of desired resources here includes the desired resources from previous runs (i.e. was desired -> created -> now observed -> now included in desired). This forces each function to track previously created desired resources in case they are no longer desired, I hope could be avoided.

Here is my experiment setup:

- I built a function the desires `n` (provider-kubernetes) objects wrapping a configmap where `n` is coming from XR spec.
- I started with n=1, and I correctly observed 1 configmap created.
- I bumped n to 3, and not correctly had 3 configmaps.
- I set n back to 1 expecting the 2 configmap to be deleted leaving 1 configmap again.

This didn't happen, and I still had 3 configmaps. Digging further, I realized that my function gets the 3 configmaps in the desired list, and even though my function only wants 1 configmap, the other two stay as desired. 

Is putting everything observed into desired of the first function io intentional? Wondering if we can avoid that and only put PT resources there in the beginning?

Reply:
I pushed https://github.com/crossplane/crossplane/pull/3465/commits/9cae4dbbb005eef964bc5841f7c3a6ddc63ddd91, which should fix this.

Reply:
Confirming that now it works as expected 👍